src/translator/CMakeLists.txt

```diff
@@ -1,3 +1,7 @@
+if (NOT COMPILE_WASM)
+    set(SERVICE "service.cpp")
+endif()
```

Comment:
@jerinphilip Adding service.cpp file just for non-wasm builds is not a good idea.

Imagine a situation where wasm fails for a use case and we want to debug it.

If we generate native and wasm artifacts from same sources, we can quickly narrow down the problem (by trying to reproduce the same use case on native) and fix it.

Including different sources for generating native and wasm artifacts leaves too much for debugging.

Reply:
That's why there are three classes and the differences are captured in classes. Test-apps exists for all three. Also, [I must give you some credits for the bad idea](https://github.com/browsermt/marian-dev/commit/05f2517f58de493d2f42236c2d23db95a9edbd8f).

I don't like this much either. The idea is to separate common things into ServiceBase and have varying methods (enqueue, stop) in the respective classes. This is probably the best compromise.  `service.cpp` doesn't duplicate anything, so if something fails it fails in one place not separately. It's really hard to debug something where code gets swapped in/out with ifdefs.

Reply:
> Also, I must give you some credits for the bad idea.

Good that you pointed it out. If you look closely, [this file](https://github.com/browsermt/marian-dev/commit/05f2517f58de493d2f42236c2d23db95a9edbd8f) is neither compiled for native builds and neither wasm builds (which is completely in line with what I said in my comment). I would appreciate it if you take some time to understand things first before giving undue credits.

> It's really hard to debug something where code gets swapped in/out with ifdefs.

I completely understand your point and let's find a solution that works well for both of us. I can propose few things here. Give me sometime.

On a side note, where are you using `WITH_PTHREADS` compile definition in the code base now? Is it not used at all?

Reply:
>  this file is neither compiled for native builds and neither wasm builds (which is completely in line with what I said in my comment).

Wrong, which is why you had to go for the same solution I applied here. I triggered the reference error where the linking failed and you had to place guards. Let's settle with s/credits/inspiration.

> On a side note, where are you using WITH_PTHREADS compile definition in the code base now? Is it not used at all?

It's not. That's why I'm proud of this solution. All thread related code is in `service.cpp` and `service.h`. When you comment out `service.cpp` from CMake files, the thread code goes away. The ifdefs path have been moved into NonThreadedService (service_base.h, service_base.cpp). I'm going to point out we're spending more time trying to disable `std::thread` from appearing in code at all because of some WASM quirks.

Reply:
> Wrong, which is why you had to go for the same solution I applied here. I triggered the reference error where the linking failed and you had to place guards. Let's settle with s/credits/inspiration.

My comments still stand true for custom marian builds. This file is only required for training. It got triggered for full blown marian builds and were caught during CI (this is what CI is for).

>  I must give you some credits for the bad idea

Crediting me for a "bad" idea and simultaneously opting it in your development? Sounds paradoxical.

> I'm going to point out we're spending more time trying to disable std::thread from appearing in code at all because of some WASM quirks.

You are free to go ahead.

Reply:
> Crediting me for a "bad" idea and simultaneously opting it in your development? Sounds paradoxical.

Again, we're finding solutions for "hiding" mutex, lock and thread from the compiler, as I understand. There aren't many "good" solutions to begin with. My rationale is you think if it's fine in custom marian compilation, it's fine here as well. If otherwise, I'd suggest you scrutinize browsermt@marian-dev to the same levels. CI doesn't catch everything unless you add more tests to cover cases.

Here's how much are in mine:
1. [single-thread](https://github.com/browsermt/bergamot-translator-tests/blob/533a5fa34cfdce662502455988b57891b7fc4c3a/tests/basic/test_bergamot_translator_app_intgemm_8bit.cpu-threads.0.sh) bergamot-translator-app covers single thread paths.
2. [multi-thread](https://github.com/browsermt/bergamot-translator-tests/blob/533a5fa34cfdce662502455988b57891b7fc4c3a/tests/basic/test_service-cli_intgemm_8bit.cpu-threads.4.sh) (service-cli): To test multi-threaded implementation on a small scale. Also lets me find out if issues are in the integration layer.
3. [WNGT20 full](https://github.com/browsermt/bergamot-translator-tests/blob/533a5fa34cfdce662502455988b57891b7fc4c3a/speed-tests/test_wngt20_perf.sh): To ensure speed isn't compromised between commits. That performance isn't affected either. Also serves good in (soft) finding deadlocks, race-conditions by adding more data and more threads. Which is how the errors in code here were caught anyway. 

Where are those of marian-dev@wasm, ie for your cmake modifications or code modifications don't break anything existing?  

> I completely understand your point and let's find a solution that works well for both of us. I can propose few things here. Give me sometime.

Let me know if you have a cleaner solution. I'll implement it even, provided you lay-out a sketch. I'm running out of ideas, I'd have to continue developing this part of the code, and I'd prefer it to be in a state (even if troubled) that I understand.

Reply:
This may seem ironic coming from me, but please be nice to each other.

Reply:
Happy to receive constructive criticism if I crossed the line anywhere above. Thanks :)

Reply:
@jerinphilip I am fine with the current changes that you have made.

I believe if I can add scripts for testing artifacts built using wasm compatible sources (the same way I could add build scripts for the same) then that would give both of us more confidence in approving each other's PRs.

One solution is:
I can run all (whichever applicable) your test scripts (that you are currently running for artifacts built with full marian/bergamot-translator sources) for artifacts built with wasm compatible sources as well. I can help there as I helped in adding build scripts for these sources.

Reply:
@abhi-agg: I'm good once the CI hits green. If the diff (which fixes failing CI) shows anything that needs a test-script rerun, I'll run locally again on var and let know of any issues in the comments.

> I can run all (of course whichever applicable) your test scripts (that you are currently running for artifacts built with full marian/bergamot-translator sources) for artifacts built with wasm compatible sources as well. I can help there as I helped in adding build scripts for these sources.

Let's move this to a separate issue, in bergamot-translator-tests. I can support a first-setup since the documentation over there is likely to be poor. Know that your extension team will have tests and can propagate any wasm breakages back as well, so it's not strictly necessary.

Reply:
> Let's move this to a separate issue, in bergamot-translator-tests. I can support a first-setup since the documentation over there is likely to be poor.

Fine with that.

> Know that your extension team will have tests and can propagate any wasm breakages back as well, so it's not strictly necessary.

Let's have this discussion in test repo.

Resolving the conversation now.